### PR TITLE
enhancement(observability): Include VRL error in log when condition execution fails

### LINKED ISSUE #8684: Output VRL error when transform condition fails
Currently, if the VRL condition provided to a `filter` transform fails to execute, it outputs this, but doesn't actually tell you why.

```
vector 0.15.0 (x86_64-apple-darwin 994d812 2021-07-16)
```

Example:

```toml
data_dir = "/var/lib/vector/"

[sources.source0]
max_length = 102400
type = "stdin"

[transforms.transform0]
inputs = ["source0"]
type = "filter"
condition = "length(parse_regex!(.kubernetes.pod_name, r'.*deployment.*')) > 0"

[sinks.sink0]
inputs = ["transform0"]
target = "stdout"
type = "console"

[sinks.sink0.encoding]
codec = "json"
```

Running:

```
echo hello | vector --config /tmp/tmp.toml
Aug 12 11:34:36.608  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info" enable_datadog_tracing=false
Aug 12 11:34:36.609  INFO vector::app: Loading configs. paths=["/tmp/tmp.toml"]
Aug 12 11:34:36.613  INFO vector::sources::stdin: Capturing STDIN.
Aug 12 11:34:36.613  INFO vector::topology: Running healthchecks.
Aug 12 11:34:36.613  INFO vector::topology: Starting source. name="source0"
Aug 12 11:34:36.614  INFO vector::topology::builder: Healthcheck: Passed.
Aug 12 11:34:36.614  INFO vector::topology: Starting transform. name="transform0"
Aug 12 11:34:36.614  INFO vector::topology: Starting sink. name="sink0"
Aug 12 11:34:36.614  INFO vector: Vector has started. debug="false" version="0.15.0" arch="x86_64" build_id="994d812 2021-07-16"
Aug 12 11:34:36.614  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
Aug 12 11:34:36.614  INFO source{component_kind="source" component_name=source0 component_type=stdin}: vector::sources::stdin: Finished sending.
Aug 12 11:34:36.614  INFO vector::shutdown: All sources have finished.
Aug 12 11:34:36.614  INFO vector: Vector has stopped.
Aug 12 11:34:36.614  WARN transform{component_kind="transform" component_name=transform0 component_type=filter}: vector::internal_events::conditions: VRL condition execution failed. internal_log_rate_secs=120
```

Here it fails because the field is undefined, but that isn't clear from the logs.

We should also update the `route` and `reduce` transforms as well as any other conditions.

### PR DESCRIPTION
**Pull request by 001wwang**
Closes #8684 

This PR adds VRL error details as an `error` field to the log emitted when a condition fails to run. 

Taking the example described in the ticket, the following log will be produced
```
ERROR vector::internal_events::conditions: VRL condition execution failed. error=function call error for "length" at (0:61): function call error for "parse_regex" at (7:60): expected "string", got "null" internal_log_rate_secs=120
```

whereas before (`vector 0.19.0`), the log would have been
```
WARN vector::internal_events::conditions: VRL condition execution failed. internal_log_rate_secs=120
```

### CODE REVIEW

**Review by jszwedko — APPROVED** [review]
Nice, that was simpler than I expected 😄

**Review by spencergilbert — APPROVED** [review]
❤️

**Review by JeanMertz — APPROVED** [review]
I agree that we should make this change, but I'm curious if there's a chance for a regression in the soaks here.

The reason for that is that the trait itself explicitly calls out that this is a hot path so allocations should be avoided. It's why I initially didn't include the error message itself.

See: https://github.com/vectordotdev/vector/blob/5203f178ee47555fe761491ddf2e835b77f64691/src/conditions/mod.rs#L16-L28

Now, I think we have bigger fish to fry when it comes to performance, and I don't think the extra allocation here is going to matter too much, especially since it only triggers if the program fails at runtime, not when it returns a boolean result.

TL;DR: I'm glad to see this changed, I doubt it'll make a measurable difference in performance, but it's good to keep this change in mind if we ever see someone complain about a degradation in performance for transforms that use VRL conditions.
